sdk/dapp-kit/src/hooks/wallet/useConnectWallet.ts

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import type { UseMutationOptions } from '@tanstack/react-query';
+import { useMutation } from '@tanstack/react-query';
+import type {
+	StandardConnectInput,
+	StandardConnectOutput,
+	WalletAccount,
+} from '@mysten/wallet-standard';
+import { useWalletContext } from '../../components/WalletProvider.js';
+import { WalletAlreadyConnectedError, WalletNotFoundError } from '../../errors/walletErrors.js';
+import {
+	getMostRecentWalletConnectionInfo,
+	setMostRecentWalletConnectionInfo,
+} from '../../utils/walletUtils.js';
+import { walletMutationKeys } from '../../constants/walletMutationKeys.js';
+
+type ConnectWalletArgs = {
+	/** The name of the wallet as defined by the wallet standard to connect to. */
+	walletName: string;
+} & StandardConnectInput;
+
+type ConnectWalletResult = StandardConnectOutput;
+
+type UseConnectWalletMutationOptions = Omit<
+	UseMutationOptions<ConnectWalletResult, Error, ConnectWalletArgs, unknown>,
+	'mutationFn'
+>;
+
+/**
+ * Mutation hook for establishing a connection to a specific wallet.
+ */
+export function useConnectWallet({
+	mutationKey,
+	...mutationOptions
+}: UseConnectWalletMutationOptions = {}) {
+	const { wallets, currentWallet, storageAdapter, storageKey, dispatch } = useWalletContext();
+
+	return useMutation({
+		mutationKey: walletMutationKeys.connectWallet(mutationKey),
+		mutationFn: async ({ walletName, ...standardConnectInput }) => {
+			if (currentWallet) {
+				throw new WalletAlreadyConnectedError(
+					currentWallet.name === walletName
+						? `The user is already connected to wallet ${walletName}.`
+						: "You must disconnect the wallet you're currently connected to before connecting to a new wallet.",
+				);
+			}
+
+			const wallet = wallets.find((wallet) => wallet.name === walletName);
+			if (!wallet) {
+				throw new WalletNotFoundError(
+					`Failed to connect to wallet with name ${walletName}. Double check that the name provided is correct and that a wallet with that name is registered.`,
+				);
+			}
+
+			dispatch({ type: 'wallet-connection-status-updated', payload: 'connecting' });
+
+			try {
+				const connectResult = await wallet.features['standard:connect'].connect(
+					standardConnectInput,
+				);
+
+				// When connecting to a wallet, we want to connect to the most recently used wallet account if
+				// that information is present. This allows for a more intuitive connection experience!
+				const mostRecentConnectionInfo = await getMostRecentWalletConnectionInfo(
```

Comment:
This feels kind of wrong because the connect mutation shouldn't really need to know this state too deeply IMO, this really is just a side effect of  autoconnection (in fact we could even model this that way, just using effects in a hook to drive this all). Could we instead make this mutation accept an optional `address` param that gets selected by default, and then autoconnect can just use that instead?

Reply:
I had a similar feeling about this when tackling autoconnect support. The only reason I did this was because I thought it might be nice to have a UX where we remember the last account you connected to even if auto-connect wasn't turned on. To your point though, a consumer dApp could just handle that themselves if we expose an optional `address` parameter.

I'll update this 😄